[Py3] Remove package containing entry points names with colons

montefra opened the pull request:
Entry points names containing colons are legit, or at least ``pkg_resources`` can deal with them without any issue.

Unfortunately, in python 3 ``pip uninstall`` choke on them when removing a package.
This is because ConfigParser supports by defaults also the "INI" format, in which the delimiter is a colon. This is not a problem for python 2.

This PR force the delimiter to be only ``=`` to fix this issue.

I haven't add a test as now. I should have to dig into the test suite (which is huge).
In theory would be enough to add an entry point somewhere called e.g.:

    entry:point = path.to.package:function

If someone would be so kind as to point me to the correct place to add it, I would be very happy to add it and check if everything is fine.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/pypa/pip/3434)
<!-- Reviewable:end -->

xavfernandez commented:
@montefra good catch.

For the test, maybe the simplest would be to write a functional test like this one:
https://github.com/pypa/pip/blob/b681665dfb3fedb220ba81933e9fe4ec3e0a10eb/tests/functional/test_uninstall.py#L35-L54

Where you can define the `setup.py` of your project with the problematic entry_point.

montefra commented:
@xavfernandez: thank you for the pointer. I should commit the test soon.

montefra commented:
I have added a test.
I've added ``pretend`` to the ``test_requires`` list to be able to run ``py.test``.
Tox run through, except for py35:

```
  docs: commands succeeded
  packaging: commands succeeded
  pep8: commands succeeded
ERROR:   py3pep8: InterpreterNotFound: python3.3
ERROR:   py26: InterpreterNotFound: python2.6
  py27: commands succeeded
ERROR:   py33: InterpreterNotFound: python3.3
  py34: commands succeeded
ERROR:   py35: commands failed
ERROR:   pypy: InterpreterNotFound: pypy
```

Python 3.5 failed with an error which is probably fault of pytest e.g.:

```
ERROR collecting tests/functional/test_completion.py

.tox/py35/lib/python3.5/site-packages/py/_path/local.py:650: in pyimport
    __import__(modname)
<frozen importlib._bootstrap>:969: in _find_and_load
    ???
<frozen importlib._bootstrap>:954: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:892: in _find_spec
    ???
<frozen importlib._bootstrap>:873: in _find_spec_legacy
    ???
.tox/py35/lib/python3.5/site-packages/_pytest/assertion/rewrite.py:137: in find_module
    source_stat, co = _rewrite_test(state, fn_pypath)
.tox/py35/lib/python3.5/site-packages/_pytest/assertion/rewrite.py:278: in _rewrite_test
    rewrite_asserts(tree)
.tox/py35/lib/python3.5/site-packages/_pytest/assertion/rewrite.py:336: in rewrite_asserts
    AssertionRewriter().run(mod)
.tox/py35/lib/python3.5/site-packages/_pytest/assertion/rewrite.py:557: in run
    new.extend(self.visit(child))
/usr/lib/python3.5/ast.py:245: in visit
    return visitor(node)
.tox/py35/lib/python3.5/site-packages/_pytest/assertion/rewrite.py:665: in visit_Assert
    top_condition, explanation = self.visit(assert_.test)
/usr/lib/python3.5/ast.py:245: in visit
    return visitor(node)
.tox/py35/lib/python3.5/site-packages/_pytest/assertion/rewrite.py:796: in visit_Compare
    left_res, left_expl = self.visit(comp.left)
/usr/lib/python3.5/ast.py:245: in visit
    return visitor(node)
.tox/py35/lib/python3.5/site-packages/_pytest/assertion/rewrite.py:647: in generic_visit
    return res, self.explanation_param(self.display(res))
.tox/py35/lib/python3.5/site-packages/_pytest/assertion/rewrite.py:584: in display
    return self.helper("saferepr", expr)
.tox/py35/lib/python3.5/site-packages/_pytest/assertion/rewrite.py:590: in helper
    return ast.Call(attr, list(args), [], None, None)
E   TypeError: Call constructor takes either 0 or 3 positional arguments
```

xavfernandez commented on tests/functional/test_uninstall.py:
```diff
@@ -154,6 +154,32 @@ def test_uninstall_overlapping_package(script, data):
+    result = script.pip('install', pkg_path)
+    result = script.pip('list')
+    assert "ep-install (0.1)" in result.stdout
+    script.pip('uninstall', 'ep_install', '-y', expect_stderr=True)
```
Is the `expect_stderr=True` needed ?

montefra replied:
Given your question, I guess no. 
Who does it work? If ``expect_stderr=False`` and there is standard error, raises an error?
In this case I remove it and recommit.

xavfernandez replied:
> If expect_stderr=False and there is standard error, raises an error?

Yes exactly, it was needed in the distutils case, because pip prints a warning in such case.

xavfernandez commented:
Yes the pytest issue is known: it is pinned to avoid this issue https://github.com/pytest-dev/pytest/issues/1083 but would need to be bumped to work on python 3.5 ...

montefra commented:
Ok. Thank you for the pytest pointer.

montefra commented:
@xavfernandez : I've removed it. The test goes fine on my system.

If you accept the PR, do I have to do something else (changelog, other file, ...)?
Do you want me to rebase the branch?

xavfernandez commented:
Well a changelog would make my life easier ^^
cf https://github.com/pypa/pip/blob/develop/CHANGES.txt

The rebase is optional since there is no conflict.

montefra commented:
@xavfernandez : I'll add an entry under ``**8.1.0 (unreleased)**`` tomorrow.

If I see that tomorrow there are conflicts, I'll rebase.

Good night

xavfernandez commented:
@montefra thanks :)

montefra commented:
@xavfernandez : done.

I hope that all is fine.

xavfernandez commented on pip/req/req_install.py:
```diff
@@ -727,7 +727,11 @@ def uninstall(self, auto_confirm=False):
+            if six.PY2:
+                options = {}
+            else:
+                options = {"delimiters": ('=')}
```
I think you meant `('=',)`.
Sorry I did not spot this before...

montefra replied:
of course... Me dumm! I'm not used to use tuples. Thank for spotting it. Committing

montefra commented:
@xavfernandez I'm about to rebase and push the commit

montefra commented:
@xavfernandez : rebase done

xavfernandez commented:
Thanks for your patience :)

montefra commented:
thank for accepting the PR and mostly thankd for pip.
